File: 2023/UICPC Regional/Round #12/solutions/autocorrect/per.py
import sys

class Trie:
    def __init__(self):
        self.aut_len = -1
        self.aut_sym = 0
        self.next = [None]*26

    def get_next(self, c, make=False):
        sym = ord(c)-ord('a')
        if make and self.next[sym] is None: self.next[sym] = Trie()
        return self.next[sym]
                

seq = [None]*(1<<20)
opt = [None]*(1<<20)

def solve(word, trie):
    L = 0
    while L < len(word) and trie is not None:
        seq[L] = trie
        trie = trie.get_next(word[L])
        L += 1

    opt[L] = len(word)-L
    nxtb = L
    for i in range(L-1, -1, -1):
        opt[i] = 1 + opt[i+1]
        if i == 0 or seq[i].aut_sym != word[i]:
            nxtb = i
        else:
            opt[i] = min(opt[i], 1 + seq[i].aut_len - (nxtb-i) + opt[nxtb])

    return opt[0];



(n,m) = map(int, sys.stdin.readline().split())
trie = Trie()
for _ in range(n):
    word = sys.stdin.readline().strip()
    node = trie
    idx = 0
    while idx < len(word):
        node = node.get_next(word[idx], make=True)
        idx += 1
        if node.aut_len == -1:
            node.aut_len = len(word)-idx
            node.aut_sym = word[idx] if idx < len(word) else None
    
for _ in range(m):
    print solve(sys.stdin.readline().strip(), trie)

File: 2023/UICPC Regional/Round #12/solutions/autocorrect/per.cc
#include <bits/stdc++.h>

struct trie {
  trie *next[26];
  int aut_len;
  char aut_sym;

  trie(): aut_len(-1), aut_sym(0) { memset(next, 0, sizeof(next)); }

  trie *get_next(char c) {
    return next[c-'a'] ? next[c-'a'] : NULL;
  }
  
  void add(const char *s) {
    if (*s) {
      if (!next[*s-'a']) next[*s-'a'] = new trie();
      next[*s-'a']->add(s+1);
    }
    if (aut_len == -1) {
      aut_len = *s ? next[*s-'a']->aut_len + 1 : 0;
      aut_sym = *s;     
    }
  }
};

char buf[1<<24];
trie *seq[1<<24];
int opt[1<<24];

int solve(const char *word, trie *t) {
  int L = 0;
  for ( ; word[L] && t; seq[L] = t, t = t->get_next(word[L++]));
  opt[L] = strlen(word)-L;
  for (int i = L-1, nxtb = L; i >= 0; --i) {
    opt[i] = 1 + opt[i+1];
    if (!i || seq[i]->aut_sym != word[i])
      nxtb = i;
    else
      opt[i] = std::min(opt[i], 1 + seq[i]->aut_len - (nxtb-i) + opt[nxtb]);
  }
  return opt[0];
}

int main(void) {
  int n, m;
  trie *t = new trie;
  scanf("%d%d", &n, &m);
  for (int i = 0; i < n; ++i) {
    scanf("%s", buf);
    t->add(buf);
  }
  for (int i = 0; i < m; ++i) {
    scanf("%s", buf);
    printf("%d\n", solve(buf, t));
  }
}
